# Aula 2

- Métodos de busca
- Métodos de ordenação

## Métodos de busca

### Busca sequencial
- Percorre sequencialmente os elementos da lista até achar ou não

In [1]:
from random import sample

# Gerar uma lista de 50 números aleatórios entre 0 e 200, sem repetições
def gerar_lista_randomica_sem_repeticoes(qtde: int):
    populacao = range(201) # range(201) vai de 0 a 200
    minha_lista = sample(populacao, qtde) # sample escolhe uma quantidade de elementos únicos aleatórios
    return minha_lista

In [2]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

[99, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 15, 166, 52, 65, 136, 187, 144, 46, 126, 38, 53, 149, 45, 69, 58, 10, 134, 40, 195, 27, 98, 90, 7, 110, 51, 145, 175, 61, 188, 94, 59, 72, 137, 85, 108, 106]


[99, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 15]

iteração 1, procurando o 15. percorro 14 posições da lista até achar:
[99, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 15]

achei o 15, troco com o primeiro da lista
[15, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 99]

iteração 2, procurando o 15. percorro 1 posições da lista até achar:
[15, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 99]

In [3]:
def busca_sequencial(lista_dados, valor_alvo):
    # Inicializa o contador de comparações
    contador_comparacoes = 0
    
    # Percorre cada item na lista
    for item in lista_dados:
        contador_comparacoes += 1  # Incrementa o contador
        
        # Verifica se o item é igual ao valor alvo
        if item == valor_alvo:
            print("contagem de comparações:", contador_comparacoes)
            return item, contador_comparacoes
    
    # Se não encontrou o valor alvo na lista
    print("contagem de comparações:", contador_comparacoes)
    return None, contador_comparacoes

In [5]:
print(minha_lista)
valor_encontrado, contagem_comparacoes = busca_sequencial(minha_lista, 106)
print(valor_encontrado)
print(contagem_comparacoes)

[99, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 15, 166, 52, 65, 136, 187, 144, 46, 126, 38, 53, 149, 45, 69, 58, 10, 134, 40, 195, 27, 98, 90, 7, 110, 51, 145, 175, 61, 188, 94, 59, 72, 137, 85, 108, 106]
contagem de comparações: 50
106
50


In [6]:
print(minha_lista)
valor_encontrado, contagem_comparacoes = busca_sequencial(minha_lista, 0)
print(valor_encontrado)
print(contagem_comparacoes)

[99, 193, 141, 157, 114, 189, 152, 101, 43, 18, 34, 130, 35, 8, 15, 166, 52, 65, 136, 187, 144, 46, 126, 38, 53, 149, 45, 69, 58, 10, 134, 40, 195, 27, 98, 90, 7, 110, 51, 145, 175, 61, 188, 94, 59, 72, 137, 85, 108, 106]
contagem de comparações: 50
None
50


In [ ]:
# Base
# FUNÇÃO busca_sequencial(lista_dados, valor_alvo):
#     DEFINIR contador_comparacoes como 0
    
#     PARA CADA item EM lista_dados:
#         INCREMENTAR contador_comparacoes em 1
        
#         SE valor_alvo É IGUAL A item:
#             IMPRIMIR "contagem de comparações: ", contador_comparacoes
#             RETORNAR item, contador_comparacoes
    
#     IMPRIMIR "contagem de comparações: ", contador_comparacoes
#     RETORNAR NULO, contador_comparacoes
# FIM FUNÇÃO
def busca_sequencial(lista_dados, valor_alvo):
    # Inicializa o contador de comparações
    contador_comparacoes = 0
    
    # Percorre cada item na lista
    for item in lista_dados:
        contador_comparacoes += 1  # Incrementa o contador
        
        # Verifica se o item é igual ao valor alvo
        if item == valor_alvo:
            print("contagem de comparações:", contador_comparacoes)
            return item, contador_comparacoes
    
    # Se não encontrou o valor alvo na lista
    print("contagem de comparações:", contador_comparacoes)
    return None, contador_comparacoes


In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

In [ ]:
# Importante: Alterar os valores que são procurados com base na lista gerada

valor_encontrado, contagem = busca_sequencial(minha_lista, 13)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial(minha_lista, 155)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial(minha_lista, 95)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial(minha_lista, 140)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial(minha_lista, 175)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial(minha_lista, 91)
print(valor_encontrado, contagem)
print()


### Busca sequencial com transposição

- Estratégia de design: *transform-and-conquer*
- Percorre sequencialmente os elementos da lista até achar ou não. Caso ache, troca a posição com o elemento da esquerda, "passando na frente" para caso ocorra uma nova busca pelo mesmo em sequência

In [ ]:
# Transpondo valores
# FUNÇÃO busca_sequencial_com_transposicao(lista_dados, valor_alvo):
#     DEFINIR contador_comparacoes como 0
    
#     PARA CADA indice, item EM ENUMERATE(lista_dados):
#         INCREMENTAR contador_comparacoes em 1
        
#         SE valor_alvo É IGUAL A item:
#             IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
            
#             SE indice É MAIOR QUE 0:
#                 DEFINIR valor_encontrado = lista_dados[indice]
#                 DEFINIR lista_dados[indice] = lista_dados[indice - 1]
#                 DEFINIR lista_dados[indice - 1] = valor_encontrado
#                 IMPRIMIR "Elemento '", valor_alvo, "' transposto de índice ", indice, " para ", indice - 1, "."
            
#             RETORNAR valor_alvo, contador_comparacoes
    
#     IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
#     RETORNAR NULO, contador_comparacoes
# FIM FUNÇÃO


In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

In [ ]:
# Importante: Alterar os valores que são procurados com base na lista gerada
print(minha_lista)

In [ ]:
# Importante: Alterar os valores que são procurados com base na lista gerada

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 13)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 155)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 95)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 140)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 175)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_transposicao(minha_lista, 91)
print(valor_encontrado, contagem)
print()


### Busca sequencial trocando valor com o primeiro da lista

- Estratégia de design: *transform-and-conquer*
- Percorre sequencialmente os elementos da lista até achar ou não. Caso ache, troca a posição com o primeiro elemento da lista e "passando na frente" para caso ocorra uma nova busca pelo mesmo em sequência

In [10]:
# Colocando em primeiro
# FUNÇÃO busca_sequencial_com_mtf(lista_dados, valor_alvo):
#     DEFINIR contador_comparacoes COMO 0
    
#     PARA CADA indice, item EM ENUMERATE(lista_dados):
#         INCREMENTAR contador_comparacoes EM 1
        
#         SE valor_alvo É IGUAL A item:
#             IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
            
#             SE indice É MAIOR QUE 0:
#                 DEFINIR valor_encontrado = REMOVER_ITEM(lista_dados, indice) // Equivalente a data_list.pop(index)
#                 INSERIR_ITEM(lista_dados, 0, valor_encontrado) // Equivalente a data_list.insert(0, found_value)
#                 IMPRIMIR "Elemento '", valor_alvo, "' movido para a primeira posição (índice 0)."
            
#             RETORNAR valor_alvo, contador_comparacoes
    
#     IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
#     RETORNAR NULO, contador_comparacoes
# FIM FUNÇÃO
# mtf = move to front
def busca_sequencial_com_mtf(lista_dados, valor_alvo):
    # Inicializa o contador de comparações
    contador_comparacoes = 0
    
    # Percorre cada item na lista
    for indice, item in enumerate(lista_dados):
        contador_comparacoes += 1  # Incrementa o contador

        # Verifica se o item é igual ao valor alvo
        if item == valor_alvo:
            if indice > 0:
                valor_encontrado = lista_dados.pop(indice) # estamos removendo o item achado da lista
                lista_dados.insert(0, valor_encontrado)
            print("contagem de comparações:", contador_comparacoes)
            return item, contador_comparacoes
    
    # Se não encontrou o valor alvo na lista
    print("contagem de comparações:", contador_comparacoes)
    return None, contador_comparacoes

In [11]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

[106, 16, 65, 80, 63, 126, 7, 32, 160, 170, 130, 60, 76, 200, 67, 136, 4, 128, 161, 146, 38, 152, 52, 153, 85, 30, 88, 89, 117, 87, 103, 74, 150, 176, 61, 47, 86, 122, 140, 194, 40, 42, 19, 105, 134, 56, 6, 147, 151, 9]


In [13]:
# Importante: Alterar os valores que são procurados com base na lista gerada

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 103)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 103)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 95)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 140)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 175)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_com_mtf(minha_lista, 91)
print(valor_encontrado, contagem)
print()


contagem de comparações: 2
103 2

contagem de comparações: 1
103 1

contagem de comparações: 50
None 50

contagem de comparações: 2
140 2

contagem de comparações: 50
None 50

contagem de comparações: 50
None 50



### Busca sequencial para listas ordenadas
- Estratégia de design: *brute force* mais inteligente
- Caso o elemento que estou na lista seja maior que o que procuro, posso encerrar a procura pois ele não está na lista.

In [ ]:
# Parando o algoritmo em caso de a lista estar ordenada e o elemento for maior que o procurado

# FUNÇÃO busca_sequencial_ordenada(lista_dados, valor_alvo):
#     DEFINIR contador_comparacoes COMO 0
    
#     PARA CADA item EM lista_dados:
#         INCREMENTAR contador_comparacoes EM 1
        
#         SE item É MAIOR QUE valor_alvo:
#             IMPRIMIR "Contagem de Comparações: ", contador_comparacoes, " (Parada antecipada: ", item, " > ", valor_alvo, ")"
#             RETORNAR NULO, contador_comparacoes // O valor não está na lista
        
#         SE item É IGUAL A valor_alvo:
#             IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
#             RETORNAR valor_alvo, contador_comparacoes
            
#     IMPRIMIR "Contagem de Comparações: ", contador_comparacoes
#     RETORNAR NULO, contador_comparacoes
# FIM FUNÇÃO

In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

In [ ]:
# Importante: Alterar os valores que são procurados com base na lista gerada
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
minha_lista.sort()
print(minha_lista)

In [ ]:
# Importante: Alterar os valores que são procurados com base na lista gerada

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 13)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 155)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 95)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 140)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 175)
print(valor_encontrado, contagem)
print()

valor_encontrado, contagem = busca_sequencial_ordenada(minha_lista, 91)
print(valor_encontrado, contagem)
print()


### Busca binária

- O array deve estar ordenado
- Reduzimos pela metade o espaço de busca a cada iteração

["Paulo Azevedo", "Pedro Bueno", "Cassio Bueno", "Miguel De Cervantes", "Luiza Dinarte", "Lucas Eduardo", "Michael Foucauldt", "Jorge Jose", "Zão Silva" ]

["Luiza Dinarte", "Lucas Eduardo", "Michael Foucauldt", "Jorge Jose", "Zão Silva" ]

["Luiza Dinarte", "Lucas Eduardo", "Michael Foucauldt"]

In [14]:
# Busca Binária
from math import floor

def busca_binaria(lista_dados, valor_alvo):
    limite_inferior = 0
    limite_superior = len(lista_dados) -1

    while limite_inferior <= limite_superior:
        meio = floor((limite_inferior + limite_superior) / 2)
        # meio = (inferior + superior) // 2
        # meio = int((inferior + superior) / 2))

        if lista_dados[meio] == valor_alvo:
            return lista_dados[meio]
        elif lista_dados[meio] < valor_alvo:
            limite_inferior = meio + 1
        else:
            limite_superior = meio - 1
    return None


# FUNÇÃO busca_binaria(lista_dados, valor_alvo):
#     DEFINIR limite_inferior = 0
#     DEFINIR limite_superior = tamanho(lista_dados) - 1
    
#     ENQUANTO limite_inferior É MENOR OU IGUAL A limite_superior:
#         DEFINIR meio = ARREDONDAR_PARA_BAIXO((limite_inferior + limite_superior) DIVIDIDO POR 2)
        
#         IMPRIMIR "inferior: ", limite_inferior, ", superior: ", limite_superior, ", meio: ", meio
        
#         SE lista_dados[meio] É IGUAL A valor_alvo:
#             RETORNAR lista_dados[meio] // O valor foi encontrado no meio
#         SENÃO SE lista_dados[meio] É MENOR QUE valor_alvo:
#             DEFINIR limite_inferior = meio + 1 // Busca na metade superior da lista
#         SENÃO: // lista_dados[meio] É MAIOR QUE valor_alvo
#             DEFINIR limite_superior = meio - 1 // Busca na metade inferior da lista
    
#     RETORNAR NULO // O valor não foi encontrado na lista
# FIM FUNÇÃO

In [15]:
minha_lista = gerar_lista_randomica_sem_repeticoes(50)
print(minha_lista)

[59, 49, 168, 98, 71, 189, 180, 6, 102, 128, 40, 164, 8, 196, 147, 108, 17, 174, 80, 117, 141, 18, 31, 92, 178, 140, 120, 65, 185, 116, 76, 165, 54, 15, 7, 89, 114, 67, 173, 197, 35, 46, 191, 200, 176, 151, 94, 47, 25, 62]


In [16]:
sorted_lista = sorted(minha_lista)
encontrado = busca_binaria(sorted_lista, 147)
print(f'Valor encontrado: {encontrado}')


Valor encontrado: 147


## Métodos de ordenação

Métodos de ordenação baseados em comparação de chaves
Bubble sort, Merge sort, Insertion Sort, Quick Sort, etc

Métodos de ordenação não baseados em comparação de chaves
Bucket sort, Radix sort

### Bubble Sort

- Estratégia de design: *brute force*
- Realiza da comparação de um elemento com todos os demais elementos da lista, e vai movendo os maiored deles para o final da lista.

Casos bons para o algoritmo:

- Listas pequenas
- Quando a performance não é necessária, por ser simples e fácil de implementar.
  

#### Exemplo de funcionamento
https://www.youtube.com/watch?v=Iv3vgjM8Pv4&ab_channel=kamalyassin
https://www.cs.usfca.edu/~galles/visualization/ComparisonSort.html

![Bubble Sort](https://upload.wikimedia.org/wikipedia/commons/c/c8/Bubble-sort-example-300px.gif)

#### Comparação Big-O
Abaixo podemos visualizar a complexidade de alguns dos algoritmos desenvolvidos:
 
Algoritmo         | melhor caso       | média               | pior caso
---------         | ------            | ------              | ------
Bubble Sort       | O(n<sup>2</sup>)  | O(n<sup>2</sup>)    | O(n<sup>2</sup>)

In [ ]:
# Bubble Sort

# FUNÇÃO bubble_sort(lista):
#     DEFINIR n = TAMANHO(lista)
    
#     PARA i DE 0 ATÉ n-1:
#         PARA j DE 0 ATÉ n-2: // n-1 para evitar índice fora do limite ao acessar lista[j+1]
#             SE lista[j] É MAIOR QUE lista[j+1]:
#                 TROCAR lista[j] COM lista[j+1]
#                 // Equivalente a:
#                 // DEFINIR temp = lista[j]
#                 // DEFINIR lista[j] = lista[j+1]
#                 // DEFINIR lista[j+1] = temp
                
#     RETORNAR lista
# FIM FUNÇÃO

In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(10)
print(minha_lista)

In [ ]:
lista_ordenada = bubble_sort(minha_lista)
print(lista_ordenada)

### Insertion Sort

- Estratégia de design: brute force
- Percorre a lista e insere o elemento na ordem correta.

Casos bons para o algoritmo:

- Lista pequena
- Listas parcialmente ordenadas

#### Exemplo de funcionamento
https://www.youtube.com/watch?v=ROalU379l3U&ab_channel=AlgoRythmics
https://www.cs.usfca.edu/~galles/visualization/ComparisonSort.html


![Insertion Sort](https://upload.wikimedia.org/wikipedia/commons/0/0f/Insertion-sort-example-300px.gif)

#### Comparação Big-O
Abaixo podemos visualizar a complexidade de alguns dos algoritmos desenvolvidos:
 
Algoritmo         | melhor caso       | média               | pior caso
---------         | ------            | ------              | ------
Bubble Sort       | O(n<sup>2</sup>)  | O(n<sup>2</sup>)    | O(n<sup>2</sup>)
Insertion Sort    | O(n)              | O(n<sup>2</sup>)    | O(n<sup>2</sup>)

In [ ]:
# Insertion Sort
# FUNÇÃO insertion_sort(minha_lista):
#     DEFINIR i = 0
#     DEFINIR j = 0
#     DEFINIR temporario = 0
    
#     ENQUANTO j É MENOR QUE TAMANHO(minha_lista):
#         DEFINIR temporario = minha_lista[j]
#         DEFINIR i = j - 1
        
#         ENQUANTO i É MAIOR OU IGUAL A 0 E minha_lista[i] É MAIOR QUE temporario:
#             DEFINIR minha_lista[i+1] = minha_lista[i]
#             DECREMENTAR i EM 1
            
#         DEFINIR minha_lista[i+1] = temporario
#         INCREMENTAR j EM 1
        
#     RETORNAR minha_lista
# FIM FUNÇÃO

In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(10)
print(minha_lista)

In [ ]:
minha_lista = insertion_sort(minha_lista)
print(minha_lista)

No melhor dos casos, quando a lista já está ordenada, ele vai percorrer uma vez só a lista de tamanho `n`, o que nos dá uma complexidade `O(n)`. No pior dos casos, em que a ordem da lista está invertida, percorreremos 2 vezes a lista `n`, e a complexidade deste algoritmo passa a ser `O(n`<sup>`2`</sup>`)`. Na média, seu desempenho é `O(n`<sup>`2`</sup>`)`.

### Merge Sort

- Estratégia de design: *divide-and-conquer*
- Realiza o ordenamento na etapa *conquer*
- Divide a entrada `n` em `k` partes menores distintas de tamanho `n/2` em que `n > k > 1`, gerando assim `k` subproblemas. Resolve (ordena) os sub problemas menores na hora de juntar os pedaços `k`, até formar a entrada de tamanho `n` ordenada.

Casos bons para o algoritmo:

- Apresenta uma boa performance em todos os casos

#### Exemplo de funcionamento
https://www.youtube.com/watch?v=XaqR3G_NVoo&t=181s&ab_channel=AlgoRythmics
https://www.cs.usfca.edu/~galles/visualization/ComparisonSort.html

![Merge Sort](https://upload.wikimedia.org/wikipedia/commons/c/cc/Merge-sort-example-300px.gif)

#### Comparação Big-O
Abaixo podemos visualizar a complexidade de alguns dos algoritmos desenvolvidos:
 
Algoritmo         | melhor caso       | média               | pior caso
---------         | ------            | ------              | ------
Bubble Sort       | O(n<sup>2</sup>)  | O(n<sup>2</sup>)    | O(n<sup>2</sup>)
Insertion Sort    | O(n)              | O(n<sup>2</sup>)    | O(n<sup>2</sup>)
Merge Sort        | O(n log n)        | O(n log n)          | O(n log n)

In [ ]:
# Merge Sort

# FUNÇÃO chamar_merge_sort(minha_lista):
#     RETORNAR merge_sort(minha_lista, 0, TAMANHO(minha_lista) - 1)
# FIM FUNÇÃO

# FUNÇÃO merge_sort(minha_lista, inferior, superior):
#     SE inferior É MENOR QUE superior:
#         DEFINIR meio = ARREDONDAR_PARA_BAIXO((inferior + superior) DIVIDIDO POR 2)
        
#         // Chamadas recursivas para dividir as sub-listas
#         CHAMAR merge_sort(minha_lista, inferior, meio)
#         CHAMAR merge_sort(minha_lista, meio + 1, superior)
        
#         // Combina as sub-listas ordenadas
#         RETORNAR unir_merge_sort(minha_lista, inferior, meio, superior)
# FIM FUNÇÃO

# FUNÇÃO unir_merge_sort(minha_lista, inferior, meio, superior):
#     DEFINIR h = inferior
#     DEFINIR i = inferior
#     DEFINIR j = meio + 1
#     DEFINIR lista_auxiliar = CRIAR_LISTA(TAMANHO(minha_lista)) // Ex: [-1] * len(my_list)
    
#     // Combina as duas sub-listas em lista_auxiliar
#     ENQUANTO h É MENOR OU IGUAL A meio E j É MENOR OU IGUAL A superior:
#         SE minha_lista[h] É MENOR OU IGUAL A minha_lista[j]:
#             DEFINIR lista_auxiliar[i] = minha_lista[h]
#             INCREMENTAR h EM 1
#         SENÃO:
#             DEFINIR lista_auxiliar[i] = minha_lista[j]
#             INCREMENTAR j EM 1
#         INCREMENTAR i EM 1
        
#     // Copia elementos restantes da primeira sub-lista, se houver
#     SE h É MAIOR QUE meio:
#         DEFINIR k = j
#         ENQUANTO k É MENOR OU IGUAL A superior:
#             DEFINIR lista_auxiliar[i] = minha_lista[k]
#             INCREMENTAR k EM 1
#             INCREMENTAR i EM 1
#     // Copia elementos restantes da segunda sub-lista, se houver
#     SENÃO:
#         DEFINIR k = h
#         ENQUANTO k É MENOR OU IGUAL A meio:
#             DEFINIR lista_auxiliar[i] = minha_lista[k]
#             INCREMENTAR k EM 1
#             INCREMENTAR i EM 1
            
#     // Copia os elementos ordenados da lista_auxiliar de volta para a minha_lista original
#     DEFINIR k = inferior
#     ENQUANTO k É MENOR OU IGUAL A superior:
#         DEFINIR minha_lista[k] = lista_auxiliar[k]
#         INCREMENTAR k EM 1
        
#     RETORNAR minha_lista
# FIM FUNÇÃO

In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(10)
print(minha_lista)

In [ ]:
minha_lista = chamar_merge_sort(minha_lista)
print(minha_lista)

Podemos observar que o algoritmo Merge Sort realiza sucessivas chamadas recursivas no array de tamanho `n`, dividindo do `índice menor` até o `meio` do array, e do `meio + 1` até o `índice maior`, até que o subproblema (chamado de `sp` mais adiante) tenha tamanho `1`,
 
Quando ele tem tamanho `1`, começa a retornar, fazendo a ordenação do `sp[indice menor, meio]` com o `sp(meio+1, indice maior)`. Para facilitar o entendimento, observe a representação do funcionamento desse algoritmo em formato de árvore, ilustrando as chamadas realizadas para um array `a` de tamanho `n = 10`

<img src="https://s3-sa-east-1.amazonaws.com/lcpi/1da35c08-ffba-4259-a396-a8b6572abb1e.png" width=600>

&nbsp;

<img src="https://i.stack.imgur.com/c0h8p.png" width=600>

### Quick Sort

- Estratégia de design: *divide-and-conquer*
- Realiza o ordenamento na etapa *divide*
- Divide a entrada `n` em `k` partes menores distintas, tal que `n > k > 1`, gerando assim `k` subproblemas. Resolve (ordena) os sub problemas menores na hora de dividir, para que quando for juntar os pedaços `k` até formar a entrada de tamanho `n`, os elementos já estejam todos ordenados.

Casos bons para o algoritmo:

- Apresenta uma boa performance em todos os casos. No pior caso, pode apresentar um desempenho pior que o MergeSort.

#### Exemplo de funcionamento
https://www.youtube.com/watch?v=ywWBy6J5gz8&ab_channel=AlgoRythmics
https://www.cs.usfca.edu/~galles/visualization/ComparisonSort.html

![Quick Sort](https://upload.wikimedia.org/wikipedia/commons/6/6a/Sorting_quicksort_anim.gif)

https://favtutor.com/resources/images/uploads/mceu_46432632011643441346270.png

#### Comparação Big-O
Abaixo podemos visualizar a complexidade de alguns dos algoritmos desenvolvidos:
 
Algoritmo         | melhor caso       | média               | pior caso
---------         | ------            | ------              | ------
Bubble Sort       | O(n<sup>2</sup>)  | O(n<sup>2</sup>)    | O(n<sup>2</sup>)
Insertion Sort    | O(n)              | O(n<sup>2</sup>)    | O(n<sup>2</sup>)
Merge Sort        | O(n log n)        | O(n log n)          | O(n log n)
Quick Sort        | O(n log n)        | O(n log n)          | O(n<sup>2</sup>)

In [ ]:
# Quicksort

# FUNÇÃO quicksort(minha_lista, inicio, fim):
#     SE inicio É MENOR QUE fim:
#         // 'pivo_indice' é a posição final do pivô após a partição
#         DEFINIR pivo_indice = dividir(minha_lista, inicio, fim)
        
#         // Chamadas recursivas para as sub-listas à esquerda e à direita do pivô
#         CHAMAR quicksort(minha_lista, inicio, pivo_indice - 1)
#         CHAMAR quicksort(minha_lista, pivo_indice + 1, fim)
        
#     RETORNAR minha_lista
# FIM FUNÇÃO

# FUNÇÃO dividir(minha_lista, inicio, fim):
#     DEFINIR pivo = minha_lista[fim] // Escolhe o último elemento como pivô
#     DEFINIR i = inicio - 1         // 'i' rastreia a posição para o próximo elemento menor que o pivô
    
#     PARA j DE inicio ATÉ fim - 1:
#         SE minha_lista[j] É MENOR OU IGUAL A pivo:
#             INCREMENTAR i EM 1
#             TROCAR minha_lista[i] COM minha_lista[j] // Move elementos menores/iguais para a esquerda
            
#     // Coloca o pivô em sua posição final correta
#     TROCAR minha_lista[i + 1] COM minha_lista[fim]
    
#     RETORNAR i + 1 // Retorna o índice onde o pivô foi colocado
# FIM FUNÇÃO

In [ ]:
minha_lista = gerar_lista_randomica_sem_repeticoes(10)
print(minha_lista)

In [ ]:
minha_lista = quicksort(minha_lista, 0, len(minha_lista) - 1)
print(minha_lista)

Ao receber um array `A`, com o indicativo do início `A[0]` e do fim do array `A[n]`, o quicksort realiza os seguintes procedimentos:
 
Seleciona um `pivo`, neste caso o elemento final `A[n]` do array que vai da posição inicial até a final (`A[0 ... n]`) e reorganiza o array de forma que os elementos à `esquerda` do `pivo` sejam menores ou iguais a ele, e a `direita` maiores ou iguais a ele. Neste momento não importa a ordem destes elementos, apenas a relação entre eles e o pivo. Esse procedimento é chamado de particionamento, e a posição final do pivo é a posição devolvida ao fim do método `Particiona`, chamada de posição `i`.
 
Após inserir o `pivo` na posição correta, realizamos uma nova chamada recursiva para o sub-array `k`, que vai da posição inicial `A[0]` até a posição `A[i - 1]` e outra chamada que vai da posição `A[i + 1]` até `A[n]`, assim sucessivamente até que atinja a condição de parada da recursão. Neste caso, é quando a posição final é maior do que a posição inicial no array (`i - 1 < 0` e `i + 1 < n`).
 
Ao final do método, o array já está ordenado, e não é necessário realizar nenhum procedimento adicional.